crypto/vrf_go.go

```diff
@@ -0,0 +1,465 @@
+
+	x.Mult32(rr2, curve25519A) // fe25519_mul(x, curve25519_A, rr2);
+	x.Negate(x)                // fe25519_neg(x, x);
+
+	x2 = &field.Element{}
+	x2.Multiply(x, x) // fe25519_sq(x2, x);
+	x3 = &field.Element{}
+	x3.Multiply(x, x2) // fe25519_mul(x3, x, x2);
+
+	e = &field.Element{}
+	e.Add(x3, x)               // fe25519_add(e, x3, x);
+	x2.Mult32(x2, curve25519A) // fe25519_mul(x2, x2, curve25519_A);
+	e.Add(x2, e)               // fe25519_add(e, x2, e);
+
+	e = chi25519(e) // chi25519(e, e);
+	s = e.Bytes()   // fe25519_tobytes(s, e);
+
+	eIsMinus1 = int(s[1] & 1) // e_is_minus_1 = s[1] & 1;
+	eIsNotMinus1 := eIsMinus1 ^ 1
+	negx = (&field.Element{}).Set(x)
+	negx.Negate(negx)                               // fe25519_neg(negx, x);
+	x.Select(x, negx, eIsNotMinus1)                 // fe25519_cmov(x, negx, e_is_minus_1);
+	x2.Zero()                                       // fe25519_0(x2);
+	x2.Select(x2, curve25519AElement, eIsNotMinus1) // fe25519_cmov(x2, curve25519_A, e_is_minus_1);
+	x.Subtract(x, x2)                               // fe25519_sub(x, x, x2);
+	// yed = (x-1)/(x+1)
+	{
+		var one, xPlusOne, xPlusOneInv, xMinusOne, yed *field.Element
+
+		one = (&field.Element{}).One()                            // fe25519_1(one);
+		xPlusOne = (&field.Element{}).Add(x, one)                 // fe25519_add(x_plus_one, x, one);
+		xMinusOne = (&field.Element{}).Subtract(x, one)           // fe25519_sub(x_minus_one, x, one);
+		xPlusOneInv = (&field.Element{}).Invert(xPlusOne)         // fe25519_invert(x_plus_one_inv, x_plus_one);
+		yed = (&field.Element{}).Multiply(xMinusOne, xPlusOneInv) // fe25519_mul(yed, x_minus_one, x_plus_one_inv);
+		s = yed.Bytes()                                           // fe25519_tobytes(s, yed);
+	}
+
+	// recover x
+	s[31] |= xSign
+
+	p3 = &edwards25519.Point{}
+	_, err := p3.SetBytes(s) // ge25519_frombytes(&p3, s) != 0
+	if err != nil {
+		// fmt.Printf("issue setting bytes: %x - %v\n", s, err)
+		return nil, err
+	}
+
+	// // multiply by the cofactor
+	p3.MultByCofactor(p3)
+
+	s = p3.Bytes() // ge25519_p3_tobytes(s, &p3);
+	return s, nil
+}
+
+func vrfNonceGeneration(truncHashedSk []byte, H *edwards25519.Point) *edwards25519.Scalar {
+	result := edwards25519.NewScalar()
+
+	hs := sha512.New()
+	hs.Write(truncHashedSk)
+	hs.Write(H.Bytes())
+	kString := hs.Sum(nil)[:64]
+	result.SetUniformBytes(kString)
+
+	return result
+}
+
+/* Subroutine specified in draft spec section 5.4.3.
+ * Hashes four points to a 16-byte string.
+ * Constant time. */
+func vrfHashPoints(P1, P2, P3, P4 *edwards25519.Point) *edwards25519.Scalar {
+	result := make([]byte, 32)
+	var str [2 + (32 * 4)]byte
+
+	str[0] = vrfSuite
+	str[1] = 0x02
+	copy(str[2+(32*0):], P1.Bytes())
+	copy(str[2+(32*1):], P2.Bytes())
+	copy(str[2+(32*2):], P3.Bytes())
+	copy(str[2+(32*3):], P4.Bytes())
+	h := sha512.New()
+	h.Write(str[:])
+	sum := h.Sum(nil)
+
+	copy(result[:], sum[:16])
+	r := edwards25519.NewScalar()
+	r.SetCanonicalBytes(result)
```

Comment:
```suggestion
	if _, err := r.SetCanonicalBytes(result); err != nil { panic(err) }
```

This will never happen, but I would rather see it checked anyway rather than leave the scalar at zero.